### PR TITLE
Per-axis quantization specification for reshape, transpose, and broadcast_in_dim.

### PR DESCRIPTION
**Pull request by sdasgup3**
Based on the discussion in https://github.com/openxla/stablehlo/issues/1574, I propose the quantization specification for the following data movement ops:
 -  **transpose**:  The `quantization_dimension` of `result` is also transposed based on the input `permutation` order.
 - **reshape**:  The `quantization_dimension` of the operand should not be merged or split.
 - **broadcast_in_dim**: For [degenerate dimension broadcasting](https://www.tensorflow.org/xla/broadcasting#broadcasting_similar-rank_arrays_with_degenerate_dimensions) of the operand's quantized dimension, the scales of the result are expanded.

### CODE REVIEW

File: docs/spec.md [hunk 1]
```diff
@@ -1384,6 +1384,12 @@ in the `operand` tensor and produces a `result` tensor. More formally,
 * (C5) For all `d` in `axes(operand)`:
   * `dim(operand, d) = 1` or
   * `dim(operand, d) = dim(result, broadcast_dimensions[d])`.
+* (C6) `is_per_tensor_quantized(operand) = is_per_tensor_quantized(result)`.
+* (C7) If `is_per_axis_quantized(result)`:
+  * `quantized_dimension(result) = broadcast_dimensions[quantized_dimension(operand)]`.
+  * `If dim(operand, quantized_dimension(operand)) = 1, then
+    scales(result)[i] = scales(operand)[0] for i in
```
Comment (loganchien) [inline docs/spec.md:1391]: `and zero_points(result)[i]] = zero_points(operand)[i]`

Comment (sdasgup3) [reply docs/spec.md]: done

File: docs/spec.md [hunk 2]
```diff
@@ -4359,20 +4365,34 @@ ordering of `index_space(result)` and `index_space(operand)`.
...
+    init_values=1, dimensions=[0], body=lambda x, y: x * y) =
+    reduce(dims(operand, [0, 1, ..., quantization_dimension(result) - 1]),
+    init_values=1, dimensions=[0], body=lambda x, y: x * y)`.
+  * `dim(operand, quantization_dimension(operand)) =
+    dim(result, quantization_dimension(result))`.
+  * `reduce(dims(operand,
+    [quantization_dimension(operand) + 1, ..., rank(operand) - 1]),
+    init_values=1, dimensions=[0], body=lambda x, y: x * y) =
+    reduce(dims(operand,
+    [quantization_dimension(operand) + 1, ..., rank(operand) - 1]),
```
[the remark this answers is not in the document]
Comment (sdasgup3) [reply docs/spec.md]: done

File: docs/spec.md [hunk 3]
```diff
@@ -1384,6 +1384,12 @@ in the `operand` tensor and produces a `result` tensor. More formally,
 * (C5) For all `d` in `axes(operand)`:
   * `dim(operand, d) = 1` or
   * `dim(operand, d) = dim(result, broadcast_dimensions[d])`.
+* (C6) `is_per_tensor_quantized(operand) = is_per_tensor_quantized(result)`.
+* (C7) If `is_per_axis_quantized(result)`:
+  * `quantized_dimension(result) = broadcast_dimensions[quantized_dimension(operand)]`.
+  * `If dim(operand, quantized_dimension(operand)) = 1, then
```
[the remark this answers is not in the document]
Comment (sdasgup3) [reply docs/spec.md]: done

File: docs/spec.md [hunk 4]
```diff
@@ -5469,22 +5490,25 @@ where `result_index[d] = operand_index[permutation[d]]`.
...
+| `result` | tensor or quantized tensor | (C1), (C3)  |
 
 #### Constraints
 
 * (C1) `element_type(operand) = element_type(result)`.
 * (C2) `permutation` is a permutation of `range(rank(operand))`.
 * (C3) `shape(result) = dim(operand, permutation...)`.
+* (C4 `is_per_tensor_quantized(operand) = is_per_tensor_quantized(result)`.
+* (C5) If `is_per_axis_quantized(result)`, then `quantized_dimension(result) =
+       permutation(quantized_dimension(operand))`.
```
Comment (loganchien) [inline docs/spec.md:5511]: Seems to be the reverse:

`quantized_dimension(operand) = permutation(quantized_dimension(result))`

because according to the code above:

```
result_index[d] = operand_index[permutation[d]]
```

`permutation` is actually a result-axis to operand-axis function (or the array index is the result-axis and the array value is the operand-axis).  so we should put `quantized_dimension(result)` in `permutation[...]`.

Comment (sdasgup3) [reply docs/spec.md]: That is good catch! Thanks. 
Updated the spec accordingly.

**Review by loganchien — APPROVED** [review]

File: docs/spec.md [hunk 5]
```diff
@@ -1384,6 +1384,13 @@ in the `operand` tensor and produces a `result` tensor. More formally,
 * (C5) For all `d` in `axes(operand)`:
   * `dim(operand, d) = 1` or
   * `dim(operand, d) = dim(result, broadcast_dimensions[d])`.
+* (C6) `is_per_tensor_quantized(operand) = is_per_tensor_quantized(result)`.
+* (C7) If `is_per_axis_quantized(result)`:
+  * `quantized_dimension(result) = broadcast_dimensions[quantized_dimension(operand)]`.
+  * If `dim(operand, quantized_dimension(operand)) = 1`, then
+    `scales(result)[i] = scales(operand)[0] and zero_points(result)[i] =
+    zero_points(operand)[0] for i in
+    range(dim(result, quantized_dimension(result)))`.
```
Comment (dansuh17) [inline docs/spec.md:1393]: totally optional - can we have an example in the spec that demonstrates the case for per-axis quantized tensor? The constraints might not be very intuitive for first-time users.

Comment (sdasgup3) [reply docs/spec.md:1393]: Hi @dansuh17 thanks for the suggestion. 
We do have examples for per-axis quantization for `uniform_quantize` and `uniform_dequantize` ([here](https://github.com/openxla/stablehlo/blob/main/docs/spec.md#uniform_quantize) and [here](https://github.com/openxla/stablehlo/blob/main/docs/spec.md#uniform_dequantize)). For most of the other ops which supports quantized types in addition to other regular element types, we follow the norm to provide example for only one regular element type (ref point 6 of [this](https://github.com/openxla/stablehlo/blob/main/docs/spec_checklist.md#stablehlo-specification-checklist)). While we add constraint checks to the verifier for each op, we will be adding example for each  constraint check. Does that sound like a plan to you?

File: docs/spec.md [hunk 6]
```diff
@@ -1384,6 +1384,13 @@ in the `operand` tensor and produces a `result` tensor. More formally,
 * (C5) For all `d` in `axes(operand)`:
   * `dim(operand, d) = 1` or
   * `dim(operand, d) = dim(result, broadcast_dimensions[d])`.
+* (C6) `is_per_tensor_quantized(operand) = is_per_tensor_quantized(result)`.
```
[the remark this answers is not in the document]
Comment (sdasgup3) [reply docs/spec.md]: Thanks for the catch! On a second thought let me remove the constraint altogether in favor of (C1) `element_type(operand) = element_type(result)` which could capture both the cases. PTAL.

**Review by doyeonkim0 — APPROVED** [review]

File: docs/spec.md (hunk 5 above)
Comment (dansuh17) [reply docs/spec.md:1392]: SGTM! Will the verifier implementation follow this commit?

**Review by dansuh17 — APPROVED** [review]

File: docs/spec.md [hunk 7]
```diff
@@ -1364,16 +1364,16 @@ in the `operand` tensor and produces a `result` tensor. More formally,
...
 #### Outputs
 
-| Name     | Type                                  | Constraints      |
-|----------|---------------------------------------|------------------|
-| `result` | tensor or per-tensor quantized tensor | (C1), (C3), (C5) |
+| Name     | Type                       | Constraints         |
+|----------|----------------------------|---------------------|
+| `result` | tensor or quantized tensor | (C1), (C3), (C5-C6) |
 
 #### Constraints
```
Comment (dansuh17) [inline docs/spec.md:1378]: Actually, following-up on Doyeon's comment at https://github.com/openxla/stablehlo/pull/1747#discussion_r1302338221, I think (C1) should be broken down into two separate conditions, one for quantized type and non-quantized type.

The reason why we need a separate issue for the quantized type is that the `element_type` is considered different when at least one of the parameter is different. If two quantized types have different quantization axes then they are considered different, e.g. `!quant.uniform<i8:f32:0, {0.1:-30,0.5:-20}> != !quant.uniform<i8:f32:1, {0.1:-30,0.5:-20}>`. If an op modifies the quantization params then C1 is violated.

I think this should be applied to all ops that **may** modify the quantization parameters.

**Review by dansuh17 — CHANGES_REQUESTED** [review]

File: docs/spec.md (hunk 5 above)
Comment (sdasgup3) [reply docs/spec.md:1397]: Yes, that is right.

**Review by paulinesho — APPROVED** [review]

File: docs/spec.md (hunk 7 above)
Comment (dansuh17) [reply docs/spec.md:1378]: Thanks! LGTM.

**Review by dansuh17 — APPROVED** [review]

File: docs/spec.md [hunk 8]
```diff
@@ -5471,20 +5498,25 @@ where `result_index[d] = operand_index[permutation[d]]`.
 
 | Label | Name          | Type                                         | Constraints |
 |-------|---------------|----------------------------------------------|-------------|
-| (I1)  | `operand`     | tensor or per-tensor quantized tensor        | (C1-C3)     |
-| (I2)  | `permutation` | 1-dimensional tensor constant of type `si64` | (C2), (C3)  |
+| (I1)  | `operand`     | tensor or quantized tensor                   | (C1-C4)     |
+| (I2)  | `permutation` | 1-dimensional tensor constant of type `si74` | (C2-C4)     |
```
Comment (dominicsymes) [inline docs/spec.md:5502]: "si74" is a typo for si64?

**Review by dominicsymes — APPROVED** [review]
Think there is one typo. Otherwise looks good to me.

File: docs/spec.md (hunk 8 above)
Comment (sdasgup3) [reply docs/spec.md]: Thanks for catching this! fixed.

File: docs/spec.md [hunk 9]
```diff
@@ -1364,26 +1364,37 @@ in the `operand` tensor and produces a `result` tensor. More formally,
...
+| `result` | tensor or quantized tensor | (C1), (C3), (C5-C6) |
 
 #### Constraints
 
-* (C1) `element_type(operand) = element_type(result)`.
+* (C1) `element_type(result)` is given by:
+  * `element_type(operand)`, if `!is_quantized(operand)`.
+  * `element_type(operand)` except that `quantization_dimension(operand)`,
+  `scales(operand)`, and `zero_points(operand)` may differ from
+  `quantization_dimension(result)`, `scales(result)`, and `zero_points(result)`
```
[the remark this answers is not in the document]
Comment (sdasgup3) [reply docs/spec.md:1385]: Yes, you are right. This should be fixed. 

Based on the Grammar rule
```
QuantizationDimension ::= IntegerConstant
```

and the fact from https://github.com/openxla/stablehlo/blob/main/docs/spec.md#names

```
For everything else, member definitions include structural parts of the program element, named after the corresponding EBNF non-terminals.
```

We can settle on `quantization_dimension`. Fixed all the occurances.

File: docs/spec.md [hunk 10]
```diff
@@ -1364,26 +1364,37 @@ in the `operand` tensor and produces a `result` tensor. More formally,
...
-| `result` | tensor or per-tensor quantized tensor | (C1), (C3), (C5) |
+| Name     | Type                       | Constraints         |
+|----------|----------------------------|---------------------|
+| `result` | tensor or quantized tensor | (C1), (C3), (C5-C6) |
 
 #### Constraints
 
-* (C1) `element_type(operand) = element_type(result)`.
+* (C1) `element_type(result)` is given by:
+  * `element_type(operand)`, if `!is_quantized(operand)`.
```
[the remark this answers is not in the document]
Comment (sdasgup3) [reply docs/spec.md]: Thanks for the catch. That is a typo to be fixed. 

This should be `!is_per_axis_quantized`. fixed.

File: docs/spec.md [hunk 11]
```diff
@@ -4361,20 +4373,36 @@ ordering of `index_space(result)` and `index_space(operand)`.
...
 #### Constraints
 
-* (C1) `element_type(operand) = element_type(result)`.
+* (C1) `element_type(result)` is given by:
+  * `element_type(operand)`, if `!is_per_axis_quantized(operand)`.
+  * `element_type(operand)` except that `quantization_dimension(operand)` and
+    `quantization_dimension(result)` may differ, otherwise.
 * (C2) `size(operand) = size(result)`.
+* (C3) If `is_per_axis_quantized(operand)`:
+  * `reduce(dims(operand, [0, 1, ..., quantization_dimension(operand) - 1]),
```
Comment (GleasonK) [inline docs/spec.md:4394]: Do we use this `reduce` notation, or something similar, elsewhere in the spec? I like that this will match the implementation fairly closely, but was a bit more difficult to understand at first glance.

I think this concern will be resolved once we have a verifier and an error message. I anticipate this error message will be a little difficult to write.. "expected dimensions of rank < quantizedDim of input and output to have the same number of elements" ?

**Review by GleasonK — APPROVED** [review]
Spoke with @sdasgup3 offline. LGTM

File: docs/spec.md (hunk 11 above)
Comment (sdasgup3) [reply docs/spec.md:4394]: Yes this notation is used in the following ops. 
[batch_norm_grad](https://github.com/openxla/stablehlo/blob/main/docs/spec.md#batch_norm_grad)
[batch_norm_training](https://github.com/openxla/stablehlo/blob/main/docs/spec.md#batch_norm_training)
[dot_general](https://github.com/openxla/stablehlo/blob/main/docs/spec.md#dot_general)
[reduce_window](https://github.com/openxla/stablehlo/blob/main/docs/spec.md#reduce_window)
[select_and_scatter](https://github.com/openxla/stablehlo/blob/main/docs/spec.md#select_and_scatter)

Or maybe something like `quantization_dimension(operand) cannot be splitted or merged during the reshape operation`
